kafka/consumer.go

```diff
@@ -207,35 +220,15 @@ func (c *Consumer) fetch(ctx context.Context) error {
 			zap.Error(err), zap.String("topic", t), zap.Int32("partition", p),
 		)
 	})
-	fetches.EachRecord(func(msg *kgo.Record) {
-		var event model.APMEvent
-		meta := make(map[string]string)
-		for _, h := range msg.Headers {
-			meta[h.Key] = string(h.Value)
-		}
-		if err := c.cfg.Decoder.Decode(msg.Value, &event); err != nil {
-			// TODO(marclop) DLQ?
-			c.cfg.Logger.Error("unable to decode message.Value into model.APMEvent",
-				zap.Error(err),
-				zap.String("topic", msg.Topic),
-				zap.ByteString("message.value", msg.Value),
-				zap.Int64("offset", msg.Offset),
-				zap.Int32("partition", msg.Partition),
-				zap.Any("headers", meta),
-			)
+	// Send partition records to be processed by its dedicated goroutine.
+	fetches.EachPartition(func(ftp kgo.FetchTopicPartition) {
+		if len(ftp.Records) == 0 {
 			return
 		}
-		ctx = queuecontext.WithMetadata(context.Background(), meta)
-		batch := model.Batch{event}
-		if err := c.cfg.Processor.ProcessBatch(ctx, &batch); err != nil {
-			c.cfg.Logger.Error("unable to process event",
-				zap.Error(err),
-				zap.String("topic", msg.Topic),
-				zap.Int64("offset", msg.Offset),
-				zap.Int32("partition", msg.Partition),
-				zap.Any("headers", meta),
-			)
-			return
+		tp := topicPartition{topic: ftp.Topic, partition: ftp.Partition}
+		select {
+		case c.consumer.consumers[tp].records <- ftp.Records:
+		case <-ctx.Done():
```

Comment:
Could you please add a comment to the `<-ctx.Done()` case about interaction with at least/most once delivery?

Reply:
done!